lib/fastimage.rb

```diff
@@ -209,77 +209,118 @@ def self.type(uri, options={})
   #   If set to true causes an exception to be raised if the image type cannot be found for any reason.
   #
   def self.animated?(uri, options={})
-    new(uri, options.merge(:animated_only=>true)).animated
+    new(uri, options.merge(:lazy=>true)).animated
   end
 
   def initialize(uri, options={})
     @uri = uri
     @options = {
+      :lazy             => false,
       :type_only        => false,
       :timeout          => DefaultTimeout,
       :raise_on_failure => false,
       :proxy            => nil,
       :http_header      => {}
     }.merge(options)
 
-    @property = if @options[:animated_only]
+    property = if @options[:animated_only]
       :animated
     elsif @options[:type_only]
       :type
     else
       :size
     end
 
-    raise BadImageURI if uri.nil?
+    @state = nil
 
-    @type, @state = nil
+    send(property) if !options[:lazy] && options[:raise_on_failure]
+  end
+
+  def type
+    fetch(:type) unless defined?(@type)
+
+    raise UnknownImageType if @options[:raise_on_failure] && !@type
+
+    @type
+  end
+
+  def size
+    begin
+      fetch(:size) unless defined?(@size)
+    rescue CannotParseImage
+    end
+
+    raise SizeNotFound if @options[:raise_on_failure] && !@size
+
+    @size
+  end
+
+  def width
+    size && size.first
+  end
+
+  def height
+    size && size.second
```

Comment:
Note that Array#second doesn't exist outside of active support, have to use size[1]